File: app/src/main/java/com/benny/openlauncher/util/DragNDropHandler.java
package com.benny.openlauncher.util;

import android.content.ClipData;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.Canvas;
import android.os.Parcelable;
import android.view.DragEvent;
import android.view.View;

import com.benny.openlauncher.activity.Home;
import com.benny.openlauncher.model.Item;
import com.benny.openlauncher.widget.AppItemView;

import org.jetbrains.annotations.Nullable;

public final class DragNDropHandler {
    private static final String DRAG_DROP_EXTRA = "DRAG_DROP_EXTRA";
    private static final String DRAG_DROP_INTENT = "DRAG_DROP_INTENT";
    public static final DragNDropHandler INSTANCE = new DragNDropHandler();
    public static Bitmap cachedDragBitmap;

    @kotlin.jvm.JvmStatic
    public static <T extends Parcelable> void startDrag(View view, Item item, DragAction.Action action, @Nullable final AppItemView.LongPressCallBack eventAction) {
        cachedDragBitmap = loadBitmapFromView(view);

        if (Home.Companion.getLauncher() != null)
            Home.Companion.getLauncher().getDragNDropView().startDragNDropOverlay(view, item, action);

        if (eventAction != null)
            eventAction.afterDrag(view);
    }

    private static Bitmap loadBitmapFromView(View view) {
        Bitmap bitmap = Bitmap.createBitmap(view.getWidth(), view.getHeight(), Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        String tempLabel = null;
        if (view instanceof AppItemView) {
            tempLabel = ((AppItemView) view).getLabel();
            ((AppItemView) view).setLabel(" ");
        }
        view.layout(0, 0, view.getWidth(), view.getHeight());
        view.draw(canvas);
        if (view instanceof AppItemView) {
            ((AppItemView) view).setLabel(tempLabel);
        }
        view.getParent().requestLayout();
        return bitmap;
    }

    public <T extends Parcelable> T getDraggedObject(DragEvent dragEvent) {

        ClipData.Item cdi = dragEvent.getClipData().getItemAt(0);
        if (cdi != null) {
            Intent intent = cdi.getIntent();
            intent.setExtrasClassLoader(Item.class.getClassLoader());
            return intent.getParcelableExtra(DRAG_DROP_EXTRA);
        }
        return null;
    }
}

File: core/src/main/java/com/benny/openlauncher/core/util/DragNDropHandler.kt
package com.benny.openlauncher.core.util

import android.content.ClipData
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Point
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import android.os.Parcelable
import android.view.DragEvent
import android.view.View
import com.benny.openlauncher.core.activity.CoreHome
import com.benny.openlauncher.core.interfaces.AbstractApp
import com.benny.openlauncher.core.manager.Setup
import com.benny.openlauncher.core.model.IconLabelItem

import com.benny.openlauncher.core.model.Item
import com.benny.openlauncher.core.viewutil.GoodDragShadowBuilder
import com.benny.openlauncher.core.widget.AppItemView

object DragNDropHandler {

    private val DRAG_DROP_EXTRA = "DRAG_DROP_EXTRA"
    private val DRAG_DROP_INTENT = "DRAG_DROP_INTENT"

    var cachedDragBitmap: Bitmap? = null

    private fun loadBitmapFromView(view: View): Bitmap {
        val bitmap = Bitmap.createBitmap(view.width, view.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        var tempLabel : String? = null
        if (view is AppItemView){
            tempLabel = view.label
            view.label = " "
        }
        view.layout(0, 0, view.width, view.height)
        view.draw(canvas)
        if (view is AppItemView){
            view.label = tempLabel
        }
        view.parent.requestLayout()
        return bitmap
    }

    @JvmStatic
    fun startDrag(view: View, item: Item, action: DragAction.Action, eventAction: AppItemView.LongPressCallBack?) {
        cachedDragBitmap = loadBitmapFromView(view)

        CoreHome.launcher?.getDragNDropView()?.startDragNDropOverlay(view, item, action)

        eventAction?.afterDrag(view)
    }

    @JvmStatic
    fun <T : Parcelable> startDrag(view: View, item: T, action: DragAction.Action, eventAction: AppItemView.LongPressCallBack?) {
        val i = Intent()
        i.putExtra(DRAG_DROP_EXTRA, item)
        val data = ClipData.newIntent(DRAG_DROP_INTENT, i)

        cachedDragBitmap = loadBitmapFromView(view)

        try {
            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                view.startDragAndDrop(data, GoodDragShadowBuilder(view), DragAction(action), 0)
            } else {
                view.startDrag(data, GoodDragShadowBuilder(view), DragAction(action), 0)
            }
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }

        eventAction?.afterDrag(view)
    }

    fun <T : Parcelable> getDraggedObject(dragEvent: DragEvent): T {
        val intent = dragEvent.clipData.getItemAt(0).intent
        intent.setExtrasClassLoader(Item::class.java.classLoader)
        return intent.getParcelableExtra(DRAG_DROP_EXTRA)
    }
}
